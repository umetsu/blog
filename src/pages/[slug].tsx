import React from 'react'
import { GetStaticPaths, GetStaticProps } from 'next'
import Layout from '../components/layout'
import { parseMarkdown } from '../utils'
import { initEnvironment } from '../graphql/relay'
import { fetchQuery } from 'relay-runtime'
import {
  ArticleQuery,
  ArticleQueryResponse,
} from '../graphql/__generated__/ArticleQuery.graphql'
import { articleQuery } from '../graphql/queries/ArticleQuery'
import { SlugsQuery } from '../graphql/__generated__/SlugsQuery.graphql'
import { slugsQuery } from '../graphql/queries/SlugsQuery'

export const getStaticPaths: GetStaticPaths = async () => {
  const environment = initEnvironment()
  const { articles } = await fetchQuery<SlugsQuery>(environment, slugsQuery, {})

  return {
    paths: articles.flatMap((a) => ({ params: { slug: a.slug } })),
    fallback: false,
  }
}

export const getStaticProps: GetStaticProps<Props, { slug: string }> = async ({
  params: { slug } = { slug: '' },
}) => {
  const environment = initEnvironment()
  const { article } = await fetchQuery<ArticleQuery>(
    environment,
    articleQuery,
    { slug }
  )
  const initialRecords = environment.getStore().getSource().toJSON()

  return {
    props: {
      article,
      initialRecords,
    },
  }
}

type Props = {
  article: ArticleQueryResponse['article']
}

function ArticleDetailPage({ article }: Props): JSX.Element {
  return (
    <Layout>
      <img src={article?.coverImage?.url} />
      <div>{article?.title}</div>
      {typeof article?.date === 'string' && <div>{article.date}</div>}
      <div>{article?.excerpt}</div>
      <div>{article?.tags}</div>
      {article?.content && (
        <div
          dangerouslySetInnerHTML={{
            __html: parseMarkdown(article.content),
          }}
        />
      )}
    </Layout>
  )
}

export default ArticleDetailPage
