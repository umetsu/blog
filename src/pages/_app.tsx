import { AppProps } from 'next/app'
import React from 'react'
import 'highlight.js/styles/vs2015.css'
import { useEnvironment } from '../graphql/relay'
import { ReactRelayContext } from 'react-relay'

function App({ Component, pageProps }: AppProps): JSX.Element {
  const environment = useEnvironment(pageProps.initialRecords)

  return (
    <ReactRelayContext.Provider value={{ environment }}>
      <Component {...pageProps} />
    </ReactRelayContext.Provider>
  )
}

export default App
