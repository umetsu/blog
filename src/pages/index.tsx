import gql from 'graphql-tag'
import { Article, ArticlesQuery } from '../graphql/generated/types'
import React from 'react'
import { GetStaticProps } from 'next'
import { fetchGraphCms } from '../graphql/server'

const articlesQuery = gql`
  query Articles {
    articles(orderBy: publishedAt_DESC) {
      id
      slug
      title
      date
    }
  }
`

export const getStaticProps: GetStaticProps<Props> = async () => {
  try {
    const {
      data: { articles },
    } = await fetchGraphCms<ArticlesQuery>(articlesQuery)
    return {
      props: {
        articles: articles,
      },
    }
  } catch (e) {
    console.error(e)
    return {
      props: {
        articles: [],
      },
    }
  }
}

type Props = {
  articles: ReadonlyArray<Pick<Article, 'id' | 'slug' | 'title' | 'date'>>
}

export function Home({ articles }: Props): JSX.Element {
  return (
    <div>
      {articles.map((article) => (
        <div key={article.id}>{article.title}</div>
      ))}
    </div>
  )
}

export default Home
