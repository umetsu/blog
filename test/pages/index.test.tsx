import React from 'react'
import { render } from '../testUtils'
import { Home } from '../../src/pages/index'

describe('Home page', () => {
  it('matches snapshot', () => {
    const { getByText } = render(
      <Home
        articles={[
          {
            id: 'id1',
            slug: 'slug1',
            title: 'Test title',
            date: new Date().getTime(),
          },
        ]}
      />,
      {}
    )
    getByText(/test title/i)
  })
})
