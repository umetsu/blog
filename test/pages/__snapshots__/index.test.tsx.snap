// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      Test title
    </div>
  </div>
</DocumentFragment>
`;
